Installation guide (#409)
Polidea/FlutterBleLib: README.md

# FlutterBleLib

A library for all your Bluetooth Low Energy needs in Flutter. Internally utilises Polidea's
[MultiPlatformBleAdapter](https://github.com/Polidea/MultiPlatformBleAdapter),
which runs on [RxAndroidBle](https://github.com/Polidea/RxAndroidBle)
and [RxBluetoothKit](https://github.com/Polidea/RxBluetoothKit).

## BLE Simulator

This library supports [BLEmulator](https://github.com/Polidea/blemulator_flutter), the BLE simulator.
The simulation allows one to develop without physical smartphone or BLE peripheral and use one's production BLE–related code in automated testing.

## Installation

To use this plugin, add `flutter_ble_lib` as a [dependency in your pubspec.yaml file](https://pub.dev/packages/flutter_ble_lib/versions/2.1.0).

### Android

> Set `minSDKVersion` in `[project]/android/app/build.gradle` file to 18.

Reviewer: It is optional but below API 18 this will not work at all

Author: If it won't work at all, then I don't see it optional. 😄

Reviewer 2: ```suggestion
Support for Bluetooth Low Energy has been added in API 18, hence the library requires `minSDKVersion` to be set to 18. If BLE is not core to your application, [you can override it](https://stackoverflow.com/questions/27335889/how-do-i-ignore-minsdkversion-of-library-in-android-studio) and handle support detection in your code: 

`[project]/android/app.build.gradle`:
```
followed by the snippet

Reviewer 2: Need to check if the link is working in MD.

# FlutterBleLib

A library for all your Bluetooth Low Energy needs in Flutter. Internally utilises Polidea's
[MultiPlatformBleAdapter](https://github.com/Polidea/MultiPlatformBleAdapter),
which runs on [RxAndroidBle](https://github.com/Polidea/RxAndroidBle)
and [RxBluetoothKit](https://github.com/Polidea/RxBluetoothKit).

## BLE Simulator

This library supports [BLEmulator](https://github.com/Polidea/blemulator_flutter), the BLE simulator.
The simulation allows one to develop without physical smartphone or BLE peripheral and use one's production BLE–related code in automated testing.

## Installation

> To use this plugin, add `flutter_ble_lib` as a [dependency in your pubspec.yaml file](https://pub.dev/packages/flutter_ble_lib/versions/2.1.0).

Reviewer 2: This is provided out of the box on pub.dev - https://pub.dev/packages/flutter_ble_lib#-installing-tab-

Reviewer 2: Ah, I see it has been only moved. It can be safely removed.

Author: Yeah, and I found that every lib I checked on pub.dev has this info in exactly this form.

Author: It's a link to this instruction. I for leaving it as it is for users who will find read this README on GitHub.

### Android

Set `minSDKVersion` in `[project]/android/app/build.gradle` file to 18.

```gradle
defaultConfig {
  ...
  // Set min SDK version to 18
  minSdkVersion 18
  targetSdkVersion 28
  ...
}
```

Add `tools:replace="android:label"` rule to `<application>` element in main `AndroidManifest.xml` file.

```xml
<application
  android:name="io.flutter.app.FlutterApplication"
  android:label="your_app"
  android:icon="@mipmap/ic_launcher"
  tools:replace="android:label">
  ...
</application>
```

Add `BLUETOOTH` and `ACCESS_FINE_LOCATION` permissions in main `AndroidManifest.xml` file.

```xml
<manifest xmlns:android="http://schemas.android.com/apk/res/android"
    xmlns:tools="http://schemas.android.com/tools"
    package="com.example.your_app">

  <uses-permission android:name="android.permission.BLUETOOTH"/>
  <uses-permission android:name="android.permission.ACCESS_FINE_LOCATION"/>

  ...
</manifest>

> ```

Reviewer 2: Permissions are mentioned in Usage section and it bloats the README. If you think it's necessary, perhaps it should be in a separate file, only linked here?

Author: Oh, yes, there is a notice about `ACCESS_FINE_LOCATION`. Then I would rather remove it from Usage section and leave it more detailed in Installation. 

For me, from a user perspective, it's more convinient to have the whole instruction inline in README instead of linked to separate file. I don't remember if I ever saw such solution on any GitHub page. As a user I always look for installation instructions on top or README and I get very frustrated when I have to search anywhere else.

Reviewer: I can create a new patch version `RxAndroidBle 1.7.1` that would have `ACCESS_FINE_LOCATION` and `android:label` removed – Manifest Merger should do the rest. This should make it easier to integrate. WDYT?

Author: The less you have to change in native project the better. I think it's a good idea.

Reviewer 2: ACCESS_FINE_LOCATION would still be required for scanning to work, and that's the only way to obtain peripherals for now.

Reviewer: But there would be only a need to have this runtime permission. No need to add it in the manifest because a dependency would have it.

Reviewer 2: Ah, nice one. Yeah, seems to be a good idea!

Reviewer: Technically I have released `1.7.1` which could be integrated in MBA and here. Should help with both manifest entries

### iOS

> In Xcode, set iOS Deployment Target (`IPHONEOS_DEPLOYMENT_TARGET`) to 10.0.

Reviewer 2: This can be lowered now I think, lemme check.

# FlutterBleLib

A library for all your Bluetooth Low Energy needs in Flutter. Internally utilises Polidea's
[MultiPlatformBleAdapter](https://github.com/Polidea/MultiPlatformBleAdapter),
which runs on [RxAndroidBle](https://github.com/Polidea/RxAndroidBle)
and [RxBluetoothKit](https://github.com/Polidea/RxBluetoothKit).

## BLE Simulator

This library supports [BLEmulator](https://github.com/Polidea/blemulator_flutter), the BLE simulator.
The simulation allows one to develop without physical smartphone or BLE peripheral and use one's production BLE–related code in automated testing.

## Installation

To use this plugin, add `flutter_ble_lib` as a [dependency in your pubspec.yaml file](https://flutter.dev/docs/development/packages-and-plugins/using-packages).

### Android

Set `minSDKVersion` in `[project]/android/app/build.gradle` file to 18.

```gradle
defaultConfig {
  ...
  // Set min SDK version to 18
  minSdkVersion 18

>   targetSdkVersion 28

Reviewer 2: targetSdkVersion should be trimmed as not related, and then the comment will be needless.

Author: I added it to give some context. But I can removed if it's not necessary.

Reviewer 2: minSdkVersion is a required property in build.gradle and I guess everyone will just do a text search for it in the file. :)

[MultiPlatformBleAdapter](https://github.com/Polidea/MultiPlatformBleAdapter),
which runs on [RxAndroidBle](https://github.com/Polidea/RxAndroidBle)
and [RxBluetoothKit](https://github.com/Polidea/RxBluetoothKit).

## BLE Simulator

This library supports [BLEmulator](https://github.com/Polidea/blemulator_flutter), the BLE simulator.
The simulation allows one to develop without physical smartphone or BLE peripheral and use one's production BLE–related code in automated testing.

## Installation

To use this plugin, add `flutter_ble_lib` as a [dependency in your pubspec.yaml file](https://flutter.dev/docs/development/packages-and-plugins/using-packages).

### Android

Set `minSDKVersion` in `[project]/android/app/build.gradle` file to 18.

```gradle
defaultConfig {
  ...
  minSdkVersion 18
  ...
}
```

Support for Bluetooth Low Energy has been added in API 18, hence the library requires `minSDKVersion` to be set to 18. If BLE is not core to your application, [you can override it](https://stackoverflow.com/questions/27335889/how-do-i-ignore-minsdkversion-of-library-in-android-studio) and handle support detection in your code.

> **Notice:** You don't need to add any additional permissions related to BLE because they are already declared in our manifest. However, you still need to request location permission at runtime, which you can implement with [permission_handler](https://pub.dev/packages/permission_handler) package. See [our example](https://github.com/Polidea/FlutterBleLib/tree/master/example) for reference.

Reviewer 2: ```suggestion
**Notice:** You don't need to add any permissions related to BLE to the AndroidManifest.xml, because they are already declared in the library's native module. However, you still need to request `ACCESS_FINE_LOCATION` permission at runtime to be able to scan for peripheral. See [example's code](https://github.com/Polidea/FlutterBleLib/blob/develop/example/lib/devices_list/devices_bloc.dart#L80) and [example's pubspec](https://github.com/Polidea/FlutterBleLib/blob/develop/example/pubspec.yaml#L20).
```

> **Notice:** You don't need to add any permissions related to BLE to the `AndroidManifest.xml`, because they are already declared in the library's native module. However, you still need to request `ACCESS_FINE_LOCATION` permission at runtime to be able to scan for peripheral. See [example's code](https://github.com/Polidea/FlutterBleLib/blob/develop/example/lib/devices_list/devices_bloc.dart#L80) and [example's pubspec](https://github.com/Polidea/FlutterBleLib/blob/develop/example/pubspec.yaml#L20).

Reviewer: We mustn't link to a specific LOC on `master` as these may change in the future. Link against a specific commit/file/LOC.

Author: Hmm... I'm not sure. If there will be a newer version of `example`, then it feels wrong to me to link to outdated code. I think that README should be update too. However, I understand that it will be hard to constantly maintain it. I don't know.

Reviewer 2: I agree with Darek. If we introduce unrelated change that doesn't outdate the example or this instruction, but moves the line inside the file, then the link will be suddenly outdated and it would be hard to spot.

> **Notice:** You don't need to add any permissions related to BLE to the `AndroidManifest.xml`, because they are already declared in the library's native module. However, you still need to request `ACCESS_FINE_LOCATION` permission at runtime to be able to scan for peripheral. See [example's code](https://github.com/Polidea/FlutterBleLib/blob/develop/example/lib/devices_list/devices_bloc.dart#L80) and [example's pubspec](https://github.com/Polidea/FlutterBleLib/blob/develop/example/pubspec.yaml#L20).

Reviewer 2: ```suggestion
**Notice:** You don't need to add any permissions related to BLE to the `AndroidManifest.xml`, because they are already declared in the library's native module. However, you still need to request `ACCESS_FINE_LOCATION` permission at runtime to be able to scan for peripheral. See [example's code](https://github.com/Polidea/FlutterBleLib/blob/202c6fe48300be207f80567ca4bee5b6fbc83eb5/example/lib/devices_list/devices_bloc.dart#L80) and [example's pubspec](https://github.com/Polidea/FlutterBleLib/blob/202c6fe48300be207f80567ca4bee5b6fbc83eb5/example/pubspec.yaml#L20).
```
